include/exadg/operators/variable_coefficients.h

```diff
@@ -235,6 +243,225 @@ class VariableCoefficients
+                             VectorType const &,
+                             Range const & face_range)
+  {
+    IntegratorFace integrator(matrix_free, true, {}, quad_index);
+
+    for(unsigned int face = face_range.first; face < face_range.second; face++)
+    {
+      integrator.reinit(face);
+
+      for(unsigned int q = 0; q < integrator.n_q_points; ++q)
+      {
+        auto const coefficient =
+          FunctionEvaluator<coefficient_rank, dim, Number>::value(coefficient_function,
+                                                                  integrator.quadrature_point(q),
+                                                                  time);
+
+        variable_coefficients.set_coefficient_face(face, q, coefficient);
+      }
+    }
+  };
+
+  /**
+   * Performs cell-based face and quadrature loops to calculate and store the variable coefficients.
+   */
+  void
+  cell_based_loop_set_coefficients(dealii::MatrixFree<dim, Number> const & matrix_free,
+                                   VectorType &,
+                                   VectorType const &,
+                                   Range const & cell_range)
+  {
+    IntegratorFace integrator(matrix_free, true, {}, quad_index);
+
+    unsigned int const n_faces_per_cell =
+      matrix_free.get_dof_handler().get_triangulation().get_reference_cells()[0].n_faces();
+
+    for(unsigned int cell = cell_range.first; cell < cell_range.second; ++cell)
+    {
+      for(unsigned int face = 0; face < n_faces_per_cell; ++face)
+      {
+        integrator.reinit(cell, face);
+
+        unsigned int const cell_based_face_index = integrator.get_current_cell_index();
+
+        for(unsigned int q = 0; q < integrator.n_q_points; ++q)
+        {
+          auto const coefficient =
+            FunctionEvaluator<coefficient_rank, dim, Number>::value(coefficient_function,
+                                                                    integrator.quadrature_point(q),
+                                                                    time);
+
+          variable_coefficients.set_coefficient_face_cell_based(cell_based_face_index,
+                                                                q,
+                                                                coefficient);
+        }
+      }
+    }
+  };
+
+  dealii::MatrixFree<dim, Number> const & matrix_free;
+  unsigned int const                      quad_index;
+  dealii::Function<dim> &                 coefficient_function;
+  double const                            time;
+  VariableCoefficients<CoefficientType> & variable_coefficients;
+};
+
+/**
+ * Visitor to access the coefficient on cells to use with variants either holding a single constant
+ * coefficient value or a VariableCoefficients object,
+ *
+ * i.e. visitor for std::variant<CoefficientType, VariableCoefficients<CoefficientType>>
+ */
+template<typename CoefficientType>
+struct GetCoefficientCell
+{
+  unsigned int const cell;
+  unsigned int const q;
```

Comment:
In terms of code design, I am not convinced that it is a good idea that the most complex case (with face, q variables) dictates the "interface" also for the simplest case of a constant coefficient (not even mentioning that it is quite hard to identify an "interface" from this struct definition). While I do not see many other variants to be realized in the future, I also do not understand the motivation for the present design if we had more variants in the future. If all these variants have different interfaces, the interface of this struct will grow according to the number of all variants, just as the simple if-elseif logic.